WIP: Add a process manager "class"

vpodzime opened the pull request:
In many places there's a need to keep track of sub-processes
spawned by the main process. Such processes need to be looked up
by various parameters, interacted with, and, eventually,
terminated.

Having a common, reliable and fast implementation with high test
coverage that holds the various pieces of information about
sub-processes together is beneficial.

Ticket: CFE-3572
Changelog: None

Steps:
- [x] initialization
- [x] adding subprocesses
- [x] getter functions
- [x] pop functions
- [x] termination functions
- [ ] fix issues on exotics
- [ ] add `void *data` to the `Subprocess` struct

vpodzime committed:
Add a process manager "class"

In many places there's a need to keep track of sub-processes
spawned by the main process. Such processes need to be looked up
by various parameters and interacted with.

Having a common, reliable and fast implementation with high test
coverage that holds the various pieces of information about
sub-processes together is beneficial.

In the future, we should extend this new API with the process
spawning and termination functionality.

Ticket: CFE-3572
Changelog: None

KotzaBoss commented on libutils/proc_manager.h:
```diff
@@ -0,0 +1,123 @@
+/**
+ * Destroy a ProcManager.
+ */
+void *ProcManagerDestroy();
```
Why is the return value of destroy `void*`?

vpodzime replied:
Good catch, should be `void` of course.

KotzaBoss commented on libutils/proc_manager.h:
```diff
@@ -0,0 +1,123 @@
+ */
+Subprocess *ProcManagerGetProcessByPID(ProcManager *manager, pid_t pid);
+Subprocess *ProcManagerGetProcessById(ProcManager *manager, const char *id);
+Subprocess *ProcManagerGetProcessByFD(ProcManager *manager, int fd);
```
could multiple processes share the same fd?
This question is shared for the other functions like pop and terminate too.

vpodzime replied:
No, FDs in one process are unique, otherwise `write(fd,...)` wouldn't know where to write to.

KotzaBoss commented on libutils/proc_manager.h:
```diff
@@ -0,0 +1,123 @@
+    pid_t pid;            /** PID of the subprocess                        [always >= 0] */
+    int in_fd;            /** FD of the input to the subprocess            [-1 if N/A] */
+    int out_fd;           /** FD of the output from the subprocess         [-1 if N/A] */
+    char lookup_fd;       /** which FD to use for lookup by FD             ['i', 'o' or '\0'] */
```
perhaps a simple enum to give a bit more context to the options?
```c
enum LookupByFD {
    IN,
    OUT,
    NONE
};

vpodzime replied:
That was my original idea too, but `enum` is an `int` in C, i.e. at least 4 bytes big. `char` is just one byte.

olehermanse replied:
If we were using C++ (C++ 11 or later) we could:

```C++
enum LookupByFD: uint8_t
{
    IN,
    OUT,
    NONE
}
```

Without it, the "workaround" is:

```C++
enum _LookupByFD
{
    IN,
    OUT,
    NONE
}

typedef LookupByFD uint8_t;
```

But I agree with your choice, this is not much "nicer".

larsewi commented on libutils/proc_manager.h:
```diff
@@ -0,0 +1,123 @@
+ */
+Subprocess *ProcManagerPopProcessByPID(ProcManager *manager, pid_t pid);
+Subprocess *ProcManagerPopProcessById(ProcManager *manager, const char *id);
+Subprocess *ProcManagerPopProcessByFD(ProcManager *manager, int fd);
```
Is not the term "pop" related to stacks? Could perhaps "remove" be better phrasing?

vpodzime replied:
I think *pop* is quite generic for the case where you *get* something while *removing* it at the same time.

vpodzime commented:
I replaced the file descriptors by `FILE` streams. Our code works with them and while it's possible to go from a `FILE` stream to an FD (`fileno()`), the other direction is impossible. Lookup functions "by FD" still make sense to me, though.

vpodzime committed:
Add MapRemoveSoft() function

Allowing values being removed from the map without being
destroyed (keys are always owned by the map and thus destroyed).

Ticket: CFE-3572
Changelog: None

vpodzime commented:
Actually, we will need the termination functionality for custom promise modules.

KotzaBoss commented on libutils/array_map.c:
```diff
@@ -86,6 +86,26 @@ bool ArrayMapRemove(ArrayMap *map, const void *key)
+
+    for (int i = 0; i < map->size; ++i)
+    {
+        if (map->equal_fn(map->values[i].key, key))
```
I looked up the declarations for the `ArrayMap` and i have a series of interlinked questions.
- Why would each `ArrayMap` "instance" need to have a pointer to the equal, destroy etc functions?
- How is it expected to be used?
- Doesnt the `ArrayMap` have _a_ specific implementation?

I saw the use of `ArrayMapNew` [here](https://github.com/cfengine/libntech/blob/828aa81863f96f243ca3b8a824e1e154f7878df1/libutils/map.c#L107) and [here](https://github.com/cfengine/libntech/blob/828aa81863f96f243ca3b8a824e1e154f7878df1/tests/unit/map_test.c#L531) but im still not sure i _exactly_ get the idea behind giving the map these three functions.

vpodzime replied:
Well, the map needs to be able to compare the keys, destroy the keys and destroy data. It's a generic key-value storage so it stores `void` pointers and needs to know how to compare them and how to destroy the data they point to.

vpodzime commented:
@olehermanse please review, if you have a chance. The only missing bit here, AFAICT, are unit tests for the termination functions.

larsewi commented on libutils/hash_map.c:
```diff
@@ -172,6 +172,40 @@ bool HashMapRemove(HashMap *map, const void *key)
+            free(cur);
+            map->load--;
+            if ((map->load < map->min_threshold) && (map->size > map->init_size))
+
```
white space

vpodzime replied:
? I don't see any white space problem in the above. Moreover, it's just a code copied from the `MapRemove()` function.

larsewi replied:
I just though an additional newline between the condition and the if-block looked kind of weird. But I guess it's alright. 👍

vpodzime replied:
Yeah, like I said copy-pasted code. But I fixed it here at least. Good catch!

vpodzime commented:
@cf-bottom jenkins, please

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=6479)](https://ci.cfengine.com//job/pr-pipeline/6479/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6479/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6479/

olehermanse commented on libutils/proc_manager.c:
```diff
@@ -0,0 +1,614 @@
+    assert(manager != NULL);
+    assert(proc != NULL);
+
+    MapRemoveSoft(manager->procs_by_pid, &(proc->pid));
+
+    if (lookup_fd != -1)
+    {
+        MapRemoveSoft(manager->procs_by_fd, &lookup_fd);
+    }
```
Wouldn't `MapRemove()` be the same, since they are using `NoopDestroy()`?

vpodzime replied:
Yes, it would. But no need to do those unnecessary function calls. :wink:

olehermanse commented on libutils/proc_manager.c:
```diff
@@ -0,0 +1,614 @@
+}
+
+
+static inline void RemoveProcFromMaps(ProcManager *manager, Subprocess *proc, int lookup_fd)
```
It seems a little backwards / error prone to me that `lookup_fd` is a parameter to this function (separate to the subprocess). `SoftRemoveProcFromMaps()` does not have this parameter.

vpodzime replied:
Because `SoftRemoveProcFromMaps()` is not used in the place where the `input/output` streams are closed already and thus unusable for `fileno()`. I'll add a comment.

olehermanse reviewed (dismissed):
Looks good in general, just some smaller comments.

vpodzime commented on libutils/array_map.c:
```diff
@@ -86,6 +86,26 @@ bool ArrayMapRemove(ArrayMap *map, const void *key)
+        {
+            map->destroy_key_fn(map->values[i].key);
+
+            memmove(map->values + i, map->values + i + 1,
+                    sizeof(MapKeyValue) * (map->size - i - 1));
```
It's just a copy-pasted code from `ArrayMapRemove()`.

vpodzime replied:
Actually, we don't have any unit tests doing `ArrayMapRemove()`. I created [CFE-3643[(https://tracker.mender.io/browse/CFE-3643) instead of trying to save the world in this PR.

vpodzime commented on libutils/proc_manager.c:
```diff
@@ -0,0 +1,614 @@
+    return proc;
+}
+
+Subprocess *ProcManagerPopProcessByFD(ProcManager *manager, int fd)
```
Yes, it's the same as `ProcManagerGetProcessByFD()` which is tested, combined with removal of the result from the internal maps which is tested by the test for `ProcManagerPopProcessByPID()`. That's good enough for me.

vpodzime commented on libutils/proc_manager.c:
```diff
@@ -0,0 +1,614 @@
+    return proc;
+}
+
+Subprocess *ProcManagerPopProcessById(ProcManager *manager, const char *id)
```
Same here.

vpodzime commented on libutils/array_map.c:
```diff
@@ -86,6 +86,26 @@ bool ArrayMapRemove(ArrayMap *map, const void *key)
+            return true;
+        }
+    }
+    return false;
```
> Your proc manager tests always use ArrayMap, never HashMap

Yeah, there are only few (fake) processes. I'll add a test with more.

vpodzime commented on libutils/proc_manager.c:
```diff
@@ -0,0 +1,614 @@
+static bool FDEqual_untyped(const void *fd1, const void *fd2)
+{
+    assert((fd1 != NULL) && (fd2 != NULL));
+    return (*((int*) fd1) == (*(int*) fd2));
```
I think it's actually clearer the way it is on the left side, saying "dereference this pointer to int" instead of "dereference this pointer type and type-cast the variable".

olehermanse replied:
So it should be changed on the right side.

vpodzime commented:
@cf-bottom jenkins with exotics, please

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=6482)](https://ci.cfengine.com//job/pr-pipeline/6482/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6482/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6482/

vpodzime commented:
There are failures on exotics. And we also need to be able to associate some custom data with the sub-processes managed.

olehermanse reviewed (approved):
Looks good to me, thanks for working on this! I see there are some remaining TODOs, but feel free to address those in separate PRs.

olehermanse reviewed (changes requested):
> There are failures on exotics. And we also need to be able to associate some custom data with the sub-processes managed.

Need to be addressed one way or another, before merging.

vpodzime commented on libutils/proc_manager.h:
```diff
@@ -0,0 +1,139 @@
+    pid_t pid;            /** PID of the subprocess                          [always >= 0]      */
+    FILE *input;          /** stream of the input to the subprocess          [can be %NULL]     */
+    FILE *output;         /** stream of the output from the subprocess       [can be %NULL]     */
+    char lookup_io;       /** which stream/FD to use for lookup by stream/FD ['i', 'o' or '\0'] */
```
This depends on the direction of the communication with the particular subprocess.

vpodzime commented on libutils/proc_manager.h:
```diff
@@ -0,0 +1,139 @@
+ * @return whether the subprocess was successfully added or not (e.g. in case of
+ *         the process with the same id, PID or lookup stream/FD added before)
+ *
+ * @note Ownership of #id, #cmd and #description is transferred to the #manager.
```
Only in case of termination.
